### PR TITLE
V2

### PR DESCRIPTION
**Pull request by fedeallocati**
I finally finished the version 2 of the implementation:
- Protocol classes that perform the creation and parsing of packets
- Individual classes for the different kind of instructions (Ping, Write, Read, Sync Write, etc), that are templated by the Protocol
- Servo classes (one for each dynamixel model), using traits to define their Control Tables, and some handy macros to write to the memory (and also to read from the memory and parse the results), both as static and as instance methods:

```
Mx64 my_mx(1);
ctrl.send(my_mx.set_goal_position(1024));
ctrl.send(Mx64::set_goal_position(1, 1024)); // Exactly same as before

ctrl.send(Mx64::get_present_position(1));
StatusPacket<Mx64::protocol_t> status_packet;
ctrl.recv(status_packet);
std::cout << Mx64::parse_present_position(status_packet) << std::endl;

```
- A base class for all the servos that contains common functionality: Read/write of the fields that all models have (e.g. model number, id, baudrate, goal position, present speed, etc) and functionality that we build on top of the common parts (for example set_goal_position_angle, that computes the exact step for the specified angle)
- Some improvement in the error handling: Specific error clases for `CRCError` (when the dynamixel sends a package and the computed CRC doesn't match the received CRC) and `StatusError` (the error bit fields in the Status Packet). They both inherit from the base class Error.
- A small change in the `Usb2Dynamixel` controller for the timeout parameter (now it is in the class instead of the `recv` method

What is missing is the adaption of the examples and tools to the new API, and some higher level functionality like search of the available servos.

I don't expect to merge rigth now, but I need feedback on this.
CC: @dogoepp @costashatz @jbmouret

### CODE REVIEW

**Comment by dogoepp** [comment]
Thanks for all that work ! I'll read and comment on the code.

For now, based on your description, what about renaming the base class for exceptions to something specific to dynamixels, like DynamixelError (or put it in the dynamixel namespace) ? I find it misleading when reading code using this library and finding exceptions instances of that type. It doesn't let the programmer know where to look for the definition of the class and what it represents.

I know you kept the existing class, the original name is not your doing. :)

**Comment by fedeallocati** [comment]
> For now, based on your description, what about renaming the base class for exceptions to something specific to dynamixels, like DynamixelError (or put it in the dynamixel namespace) ? I find it misleading when reading code using this library and finding exceptions instances of that type. It doesn't let the programmer know where to look for the definition of the class and what it represents.

Actually they are inside the `dynamixel` namespace, and also the new `errors` namespace.

I will also add some comments, so it's clear how to add new models/fix current ones.

**Comment by fedeallocati** [comment]
I finished the object oriented API.
We now have a `dynamixel::auto_detect` method (in the misc file), with inheritance and model abstraction.
With this I think that the V2 is completed and ready to merge. The arm compilation is missing(because I updated the waf version, and their API changed), and also all of the old examples, but we can add that later.

**Comment by dogoepp** [comment]
In the code, there are many assertions. Are they the proper use cases for assertions ?

cc: @jbmouret @costashatz

**Comment by costashatz** [comment]
> In the code, there are many assertions. Are they the proper use cases for assertions ?
> 
> cc: @jbmouret @costashatz

I will have a look this week... I am not convinced yet that all these assertions should be there.

**Comment by fedeallocati** [comment]
They should'nt be, but it was the quickest way to leave a TODO for throwing
exceptions :P

2016-03-01 10:58 GMT+01:00 Konstantinos Chatzilygeroudis <
notifications@github.com>:

**Comment by dogoepp** [comment]
It seems that we love typedefs. Then, why don't we use one for the packet's parameters (like [here](https://github.com/resibots/libdynamixel/pull/6/files#diff-6060f3580334ede77bbf3c80f6b8478aR53)) ? We already use typedefs for `instr_t`, even though it maps to the same type for protocol 1 and 2.

**Comment by dogoepp** [comment]
What is the difference between a reg_goal_position_angle and a set_goal_position_angle ?

File: src/dynamixel/servos/servo.hpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    } \
+    \
+    static inline InstructionPacket<typename Servo<Model>::protocol_t> reg_##Name(typename Servo<Model>::protocol_t::id_t id, typename Servo<Model>::ct_t::Name##_t value) \
+    { \
+            return typename Servo<Model>::reg_write_t(id, Servo<Model>::ct_t::Name, Servo<Model>::protocol_t::pack_data(value)); \
+    } \
+    \
+    inline InstructionPacket<typename Servo<Model>::protocol_t> set_##Name(typename Servo<Model>::ct_t::Name##_t value)  const \
+    { \
+            return Model::set_##Name(this->_id, value); \
```
Comment (dogoepp) [inline src/dynamixel/servos/servo.hpp]: Which method is this calling ?

**Comment by fedeallocati** [comment]
Reg methods store the value in a temporary memory of the servos, that gets
written to the RAM or ROM when you send it an Action instruction.
That's what we are using in dynamixel_control_hw to send the joint
positions all at the same time: we first register the positions in all the
servos, and then broadcast an Action instruction

**Comment by dogoepp** [comment]
OK thanks

**Comment by dogoepp** [comment]
As soon as all previous comments are solved, I'm ok for merging.

File: src/dynamixel/servos/servo.hpp (hunk 1 above)
Comment (fedeallocati) [inline src/dynamixel/servos/servo.hpp]: I'm not sure that I understand your question.
We always have 2 versions of a method:
The static version, accesed trough the model, e.g. `Mx28::set_moving_speed(id, speed)`,
and the instance version, which calls the static version of the same function, but passing the stored id.

Comment (dogoepp) [inline src/dynamixel/servos/servo.hpp]: You answered the question orally, thanks.

**Comment by dogoepp** [comment]
@costashatz Did you have a look at the asserts ? Shall I do this ?

**Comment by costashatz** [comment]
@dogoepp I will do it this weekend.  I would have done it today but I got
held up by a bug in hexapod_dart..

**Comment by costashatz** [comment]
> @costashatz Did you have a look at the asserts ? Shall I do this ?

@dogoepp @fedeallocati please look at #11 ..

**Comment by dogoepp** [comment]
Please have a look at #13.

Besides, ready to merge ?

**Comment by costashatz** [comment]
> Please have a look at #13.
> 
> Besides, ready to merge ?

Once we merge #13, I will format the code. Then we are ready to merge..

**Comment by dogoepp** [comment]
When do you think you coud take care of the formating @costashatz ? Do you need help on this ?

**Comment by dogoepp** [comment]
OK to merge ?

**Comment by costashatz** [comment]
> OK to merge ?

There are conflicts...

**Comment by dogoepp** [comment]
Because I updated v1 but the related file is deleted in v2 (utility 
program).

**Comment by costashatz** [comment]
> Because I updated v1 but the related file is deleted in v2 (utility
> program).

Yes but we need to fix it..

**Pull request merged**